k8s-scheduler/src/test/java/com/vmware/dcm/ScopeTest.java

```diff
@@ -0,0 +1,263 @@
+            // more spare capacity as i increases
+            node.getStatus().getCapacity().put("cpu", new Quantity(String.valueOf(10 + i)));
+            node.getStatus().getCapacity().put("memory", new Quantity(String.valueOf(1000)));
+            node.getStatus().getCapacity().put("pods", new Quantity(String.valueOf(100)));
+            nodeResourceEventHandler.onAddSync(node);
+        }
+        for (int i = 0; i < numPods; i++) {
+            handler.onAddSync(newPod("p" + i));
+        }
+
+        Model model = buildModel(conn);
+        ScopedModel scopedModel = new ScopedModel(conn, model);
+        Set<String> scopedNodes = scopedModel.getScopedNodes();
+
+        assertEquals(numPods, scopedNodes.size());
+        scopedNodes.forEach(n -> assertTrue(Integer.parseInt(n.substring(1)) >= numNodes - numPods));
+    }
+
+    /*
+     * E2E test with scheduler:
+     * Test if Scope limits candidate nodes according to spare resources
+     */
+    @Test
+    public void testScopedSchedulerSimple() {
+        final DBConnectionPool dbConnectionPool = new DBConnectionPool();
+        final DSLContext conn = dbConnectionPool.getConnectionToDb();
+        final List<String> policies = Policies.getDefaultPolicies();
+        final NodeResourceEventHandler nodeResourceEventHandler = new NodeResourceEventHandler(dbConnectionPool);
+        final int numNodes = 100;
+        final int numPods = 40;
+        conn.insertInto(Tables.BATCH_SIZE).values(numPods).execute();
+
+        final PodEventsToDatabase eventHandler = new PodEventsToDatabase(dbConnectionPool);
+        final PodResourceEventHandler handler = new PodResourceEventHandler(eventHandler::handle);
+
+        for (int i = 0; i < numNodes; i++) {
+            final Node node = newNode("n" + i, Collections.emptyMap(), Collections.emptyList());
+            // more spare capacity as i increases
+            node.getStatus().getCapacity().put("cpu", new Quantity(String.valueOf(10)));
+            node.getStatus().getCapacity().put("memory", new Quantity(String.valueOf(1000 + 1000 * i / numNodes)));
+            node.getStatus().getCapacity().put("pods", new Quantity(String.valueOf(100)));
+            nodeResourceEventHandler.onAddSync(node);
+
+            // Add one system pod per node
+            final String podName = "system-pod-n" + i;
+            final Pod pod = newPod(podName, "Running");
+            pod.getSpec().setNodeName("n" + i);
+            handler.onAddSync(pod);
+        }
+
+        for (int i = 0; i < numPods; i++) {
+            handler.onAddSync(newPod("p" + i));
+        }
+
+        final Scheduler scheduler = new Scheduler(dbConnectionPool, policies, "ORTOOLS", true, NUM_THREADS);
+        scheduler.setScopeOn();
+        scheduler.scheduleAllPendingPods(new EmulatedPodToNodeBinder(dbConnectionPool));
+
+        // Check that all pods have been scheduled to a node eligible by the scope filtering
```

Comment:
Update this comment to say that the capacities have been assigned in a certain order and that's what you're asserting on below.